syncstorage/views/authentication.py

```diff
@@ -70,46 +76,142 @@ def decode_hawk_id(self, request, tokenid):
                 # it falls within the allowable expired-token window.
                 try:
-                    data = tm.parse_token(tokenid, now=now)
+                    data = self._parse_token(tm, tokenid, now)
                     userid = data["uid"]
                 except tokenlib.errors.ExpiredTokenError:
                     recently = now - self.expired_token_timeout
-                    data = tm.parse_token(tokenid, now=recently)
+                    data = self._parse_token(tm, tokenid, recently)
                     # We replace the uid with a special string to ensure that
                     # calling code doesn't accidentally treat the token as
                     # valid. If it wants to use the expired uid, it will have
                     # to explicitly dig it back out from `request.user`.
-                    userid = data["uid"]
-                    request.user["expired_uid"] = data["expired_uid"] = userid
-                    userid = data["uid"] = "expired:%d" % (userid,)
-            except ValueError:
-                # Token validation failed, move on to the next secret.
+                    data["expired_uid"] = data["uid"]
+                    userid = data["uid"] = "expired:%d" % (data["uid"],)
+            except tokenlib.errors.InvalidSignatureError, e:
+                # Token signature check failed, try the next secret.
                 continue
+            except TypeError, e:
+                # Something went wrong when validating the contained data.
+                raise ValueError(str(e))
             else:
-                # Token validation succeeded, quit the loop.
+                # Token signature check succeeded, quit the loop.
                 break
         else:
             # The token failed to validate using any secret.
             logger.warn("Authentication Failed: invalid hawk id")
             raise ValueError("invalid Hawk id")
-        # Sanity-check the contained data.
-        # Any errors raise ValueError, triggering auth failure.
-        try:
-            token_node_name = data["node"]
-        except KeyError, e:
-            msg = "missing value in token data: %s"
-            raise ValueError(msg % (e,))
-        if token_node_name != node_name:
+
+        # Let the app access all user data from the token.
+        request.user.update(data)
+        request.metrics["metrics_uid"] = data.get("hashed_fxa_uid")
+        request.metrics["metrics_device_id"] = data.get("hashed_device_id")
+
+        # Sanity-check that we're on the right node.
+        if data["node"] != node_name:
             msg = "incorrect node for this token: %s"
-            raise ValueError(msg % (token_node_name,))
+            raise ValueError(msg % (data["node"],))
+
         # Calculate the matching request-signing secret.
         key = tokenlib.get_derived_secret(tokenid, secret=secret)
 
-        request.metrics["fxa_uid"] = data.get("fxa_uid")
```

Comment:
I assume backend systems relying on the id's are updated somewhere?

Reply:
We could only find the one place that this is being used, and it doesn't sound like anyone's using it; we'll either update or disable it in https://bugzilla.mozilla.org/show_bug.cgi?id=1500830